collectors/debugfs.plugin/sys_devices_virtual_powercap.c

```diff
@@ -0,0 +1,246 @@
+}
+
+static void
+look_for_zones(const char *control_type, struct zone_t *parent, const char *path, struct zone_t **zones, size_t *count)
+{
+    DIR *dir = opendir(path);
+    if (unlikely(dir == NULL)) {
+        return;
+    }
+
+    struct dirent *de = NULL;
+    while ((de = readdir(dir))) {
+        if (de->d_type != DT_DIR || de->d_name[0] == '.') {
+            continue;
+        }
+
+        char zone_path[FILENAME_MAX + 1];
+        snprintfz(zone_path, FILENAME_MAX, "%s/%s", path, de->d_name);
+
+        collector_info("Looking for zone in \"%s\"", zone_path);
+
+        struct zone_t *zone = get_zone(control_type, parent, zone_path, zones, count);
+        if (zone != NULL) {
+            look_for_zones(control_type, zone, zone_path, zones, count);
+        }
+    }
+
+    closedir(dir);
+}
+
+static int get_rapl_zones(struct zone_t **zones)
+{
+    *zones = NULL;
+
+    char dirname[FILENAME_MAX + 1];
+    snprintfz(dirname, FILENAME_MAX, "%s%s", netdata_configured_host_prefix, "/sys/devices/virtual/powercap");
+
+    DIR *dir = opendir(dirname);
+    if (unlikely(dir == NULL)) {
+        return 0;
+    }
+
+    size_t count = 0;
+
+    struct dirent *de = NULL;
+    while ((de = readdir(dir))) {
+        if (de->d_type != DT_DIR || de->d_name[0] == '.') {
+            continue;
+        }
+
+        char control_type_path[FILENAME_MAX + 1];
+        snprintfz(control_type_path, FILENAME_MAX, "%s/%s", dirname, de->d_name);
+
+        collector_info("Looking at control type \"%s\"", de->d_name);
+        look_for_zones(de->d_name, NULL, control_type_path, zones, &count);
+    }
+    closedir(dir);
+
+    return count;
+}
+
+static void send_chart(int update_every, const char *name)
+{
+    fprintf(
+        stdout,
+        "CHART %s '' 'Power Consumption' 'Watts' 'power consumption' '' '%s' %d %d '%s' 'debugfs.plugin' ''\n",
+        METRIC_ID,
+        debugfs_rrdset_type_name(RRDSET_TYPE_LINE),
+        NETDATA_CHART_PRIO_POWERCAP,
+        update_every,
+        name);
+}
+
+static void send_dimension(const char *zone_name)
+{
+    fprintf(
+        stdout,
+        "DIMENSION '%s' '%s' %s 1 %d ''\n",
+        zone_name,
+        zone_name,
+        debugfs_rrd_algorithm_name(RRD_ALGORITHM_ABSOLUTE),
+        1000);
+}
+
+static void send_begin()
+{
+    fprintf(stdout, "BEGIN %s\n", METRIC_ID);
+}
+
+static void send_set(const char *zone_name, collected_number value)
+{
+    fprintf(stdout, "SET %s = %lld\n", zone_name, value);
+}
+
+static void send_end_and_flush()
+{
+    fprintf(stdout, "END\n");
+    fflush(stdout);
+}
+
+int do_sys_devices_virtual_powercap(int update_every, const char *name)
+{
+    if (unlikely(g_zones_count <= 0)) {
+        g_zones_count = get_rapl_zones(&g_zones);
+        if (unlikely(g_zones_count < 0)) {
+            collector_error("Failed to find powercap zones.");
+            return 1;
+        } else if (unlikely(g_zones_count == 0)) {
+            collector_info("No powercap zones found.");
```

Comment:
Should we not use `collector_error` here too?

Reply:
It's a valid case for no power zones to be found on some systems so I don't think it's an error.

Reply:
Thank you to clarify!